plugins/glustershd/actor.go

```diff
@@ -19,69 +21,148 @@ const (
 
 type shdActor struct{}
 
-func (actor *shdActor) Do(v *volume.Volinfo, key string, value string, logger log.FieldLogger) error {
-	if key != selfHealKey {
+func (actor *shdActor) Do(v *volume.Volinfo, key string, value string, volOp xlator.VolumeOpType, logger log.FieldLogger) error {
+	voltype := v.Type.String()
+
+	if voltype != "Replicate" && voltype != "Disperse" {
 		return nil
 	}
+
 	glustershDaemon, err := newGlustershd()
 	if err != nil {
 		return err
 	}
-
-	switch value {
-	case "on":
-		if isHealEnabled(v) {
-			err = daemon.Start(glustershDaemon, true, logger)
-			if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+	switch volOp {
+	case xlator.VolumeStart:
+		for k, v := range v.Options {
+			_, _, key, err := options.SplitKey(k)
+			if err != nil {
 				return err
 			}
+			if key == selfHealKey && v == "off" {
+				return nil
+			}
 		}
-	case "off":
-		if !isHealEnabled(v) {
-			isVolRunning, err := volume.AreReplicateVolumesRunning()
-			if err != nil {
+		err := daemon.Start(glustershDaemon, true, logger)
+		if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+			return err
+		}
+
+	case xlator.VolumeStop:
+		isVolRunning, err := volume.AreReplicateVolumesRunning(v.ID)
+		if err != nil {
+			return err
+		}
+
```

Comment:
There is one SHD per node. When multiple volumes are present, we'll have to restart SHD when any one volume is stopped. And that restart should fetch the new volfile which doesn't contain the stopped volume details. This needs more thinking.

@aravindavk Thoughts ?

Reply:
today I opened an issue for the same. We need to regenerate cluster volfiles on volume start and stop(Notify not required). when these daemons restarted they will get latest volfiles automatically.

https://github.com/gluster/glusterd2/issues/1005

Reply:
I'm gonna take the liberty move this out of the current sprint until volgen changes get in.

Reply:
PRs are not related. Side effect of not having volgen patch and having only this patch is cluster level daemons will do extra work even for the stopped volumes. (Like crawl and other book keeping activities). We can review this PR and take it up since no changes required to this part of code when #1005 is addressed.